xapian-letor/ranker/ranker.cc

```diff
@@ -351,19 +351,15 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
     return firstfv.get_label() > secondfv.get_label();
 }
 
-std::vector<Xapian::docid>
-Ranker::rank(const Xapian::MSet & mset, const string & model_key, const Xapian::FeatureList & flist)
+void
+Ranker::rank(Xapian::MSet & mset, const string & model_key, const Xapian::FeatureList & flist)
 {
     LOGCALL(API, std::vector<Xapian::docid>, "Ranker::rank", mset | model_key | flist);
     std::vector<FeatureVector> fvv = flist.create_feature_vectors(mset, letor_query, Xapian::Database(db_path));
     load_model_from_metadata(model_key);
     std::vector<FeatureVector> rankedfvv = rank_fvv(fvv);
-
-    std::vector<Xapian::docid> rankeddid;
-    for (size_t i = 0; i < rankedfvv.size(); ++i) {
-	rankeddid.push_back(rankedfvv[i].get_did());
-    }
-    return rankeddid;
+    mset.replace_weights(rankedfvv.begin(), rankedfvv.end());
```

Comment:
As noted above, this requires that anyone using the API to implement iteration over objects providing a `get_score()` method.  But avoiding the temporary vector just to hold the weights is good.

Better would be for the new xapian-core API to just dereference the iterator to get the weight, and have a "shim" iterator class here which does the dereferencing.  Something like:

```
class ScoreIterator {
    std::vector<FeatureVector>::const_iterator it;

  public:
    explicit ScoreIterator(const std::vector<FeatureVector>::const_iterator it_) : it(it_) { }
 
    double operator*() const { return it->get_score(); }

    ScoreIterator& operator++() { ++it; return this; }

    bool operator==(const ScoreIterator& o) { return it == o.it; }

    // forward other methods to "it"...
};
```

Although it may look like this involves extra code and extra function calls, because it's all inlined via the template, the extra stuff just optimises away.

Being able to write simple shims like the above is the reason why it's good to avoid unnecessary extra requirements on the iterators.